### DISCUSSION TITLE
clarify use of "bomlink" in schema

### DISCUSSION THREAD

**Discussion by jkowalleck** [CycloneDX/specification#232] [Q&A]
I am still uncertain which elements allow the use of [bomlink](https://cyclonedx.org/capabilities/bomlink). 
From use cases it appears to be possible in `externalReferences[].url`, `vulnerabilities[].affects[].ref`.
Any where else? 

I'd suggest enhancing the schema to make it visible where a bomlink is allowed and where it is not.
This helps tool builders, parsers, and most importantly clarifies this from schema, not only some additional pamphlets/texts.

**Comment by stevespringett** [comment]
Related to #179

**Comment by stevespringett** [comment]
`vulnerabilities[].affects[].ref` is just a string, not a URI, but external references are entirely possible (and recommended)

**Comment by jkowalleck** [comment]
the thing is, we introduced an extra type alias "refType" for strings, to make it clear that some strings are not alike. 

- refType Identifier-DataType for interlinked elements.

in `dependencies` these the attribute `ref` are of "refType", in components the `bom-ref` is of "refType". the `dependencies[].ref` there is expected to have a counterpart in `components[].bom-ref`
no bomlink allowed.

the `vulnerabilities[].affects[].ref` is of "refType", and - AFAIK - allowed to reference a `components[].bom-ref`, or be a bomlink - which should have a certain format. 
So, `vulnerabilities[].affects[].ref` value can have a counterpart in the current CDX document, nor in another CDX document if it is in the format of a bomlink. This is not clear from the schema,


i propose to introduce a new type "bomlink", a string -alias with a constraint: it must match the pattern  
`urn:cdx:[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/[1-9][0-9*]#.+`.
1. "urn:cdx:" is hard
2. pattern from [serialnumber-uuid-part](https://github.com/CycloneDX/specification/blob/ccbf7b5781ef534cd62616e3c4221004c7c82a66/schema/bom-1.4.schema.json#LL39C29-L39C89)
3. "/" is hard
4. pattern for integer
5. "#" is hard
6. any non-empty string

This new type "bomlink" should be added wherever a "bomlink" is allowed/expected, so that digesting software authors are aware from reading the schema, that this special string at certain structures can have more meaning.
Otherwise, we would end up with lots of text describing specialties.


I might _draft_ a pullrequest to show how this might look like.

**Reply by jkowalleck** [reply]
caused #236

**Comment by jkowalleck** [comment]
@stevespringett wrote here: https://github.com/CycloneDX/specification/issues/229#issuecomment-1562170044
>  Every URL in CycloneDX can either be a URL or a [BOM-Link](https://cyclonedx.org/capabilities/bomlink/).

**Comment by jkowalleck** [comment]
Possibly caused #136
